### PR TITLE
u256 shifts operators

### PR DESCRIPTION
**Pull request by xunilrj**
## Description

This PR is part of https://github.com/FuelLabs/sway/pull/4794.

It implements left and right shifts. These have a little difference because the `rhs` is always `u64`, which means that the WQOP used the `rhs` is not indirect.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_pass/language/u256/src/main.sw [hunk 1]
```diff
@@ -26,6 +26,12 @@ fn not_operator() -> u256 {
     !(a - b)
 }
 
+// returns 8
+fn shift_operators() -> u256 {
```
Comment (anton-trunov) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/u256/src/main.sw:30]: maybe add some corner cases here too? like shifting all bits out of a `u256`, making the least significant bit the most significant and vice versa

Comment (xunilrj) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/u256/src/main.sw:30]: Done,

Comment (anton-trunov) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/u256/src/main.sw:30]: Ah, my bad, I should've used better wording. What I meant is something like `0x0...01 << 255` becomes `0x10....0`, and `0x10..0 >> 255` becomes `0x0...01`

Comment (anton-trunov) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/u256/src/main.sw:30]: And for the "shifting out" all bits `0x1...11 << 256` becomes `0x0`, and `0x11..1 >> 256` becomes `0x0` as well

Comment (xunilrj) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/u256/src/main.sw:30]: Ah... Done.

File: sway-core/src/asm_lang/virtual_immediate.rs [hunk 2]
```diff
@@ -11,6 +11,8 @@ pub enum WideOperations {
     Add = 0,
     Sub = 1,
     Not = 2,
+    Lsh = 6,
```
Comment (IGI-111) [inline sway-core/src/asm_lang/virtual_immediate.rs:14]: I assume there's an explanation for the jump in value, but what is it?

File: sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs [hunk 3]
```diff
@@ -885,9 +885,10 @@ fn type_check_shift_binary_op(
         handler,
         ctx.by_ref()
             .with_help_text("Incorrect argument type")
-            .with_type_annotation(return_type),
+            .with_type_annotation(engines.te().insert(engines, TypeInfo::Numeric)),
```
Comment (IGI-111) [inline sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs:888]: This doesn't look right, we should be type checking the intrinsics. With this code you could have `let _: u32 = 1u64 << 1u64;` which is incorrect.

What's the reason to relax it here?

File: sway-ir/src/verify.rs [hunk 4]
```diff
@@ -389,11 +389,24 @@ impl<'a, 'eng> InstructionVerifier<'a, 'eng> {
...
+            // Shifts rhs are 64 bits
+            BinaryOpKind::Lsh | BinaryOpKind::Rsh => {
+                if !arg1_ty.is_ptr(self.context)
+                    || !arg2_ty.is_uint64(self.context)
+                    || !result_ty.is_ptr(self.context)
+                {
+                    return Err(IrError::VerifyBinaryOpIncorrectArgType);
+                }
+            }
+            _ => {
```
Comment (IGI-111) [inline sway-ir/src/verify.rs:402]: You should explicit the other ops so we don't introduce a bug if a new one is added.

File: sway-ir/src/verify.rs [hunk 5]
```diff
@@ -431,8 +444,19 @@ impl<'a, 'eng> InstructionVerifier<'a, 'eng> {
...
-            return Err(IrError::VerifyBinaryOpIncorrectArgType);
+
+        match op {
+            // Shifts can have the rhs with different type
+            BinaryOpKind::Lsh | BinaryOpKind::Rsh => {
+                if !arg1_ty.is_uint(self.context) || !arg2_ty.is_uint(self.context) {
+                    return Err(IrError::VerifyBinaryOpIncorrectArgType);
+                }
+            }
+            _ => {
```
Comment (IGI-111) [inline sway-ir/src/verify.rs:455]: Same here

File: sway-core/src/asm_lang/virtual_immediate.rs (hunk 2 above)
Comment (xunilrj) [reply sway-core/src/asm_lang/virtual_immediate.rs:14]: This is coming from: http://specs.fuel.network/master/fuel-vm/instruction-set.html#wdop-misc-128-bit-integer-operations

File: sway-ir/src/verify.rs (hunk 4 above)
Comment (xunilrj) [reply sway-ir/src/verify.rs:402]: Good catch. Done.

File: sway-ir/src/verify.rs (hunk 5 above)
Comment (xunilrj) [reply sway-ir/src/verify.rs:455]: Done.

File: sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs (hunk 3 above)
Comment (xunilrj) [reply sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs]: You are right, This is wrong. Fixed.

**Review by IGI-111 — APPROVED** [review]
